### PR TITLE
Add support for hclencoder and BQ Dataset

### LINKED ISSUE #149: Add hcl marshal template funcs

### LINKED ISSUE #151: Add support for bigquery

### PR DESCRIPTION
**Pull request by umairidris**
Fix #151 
Fix #149

### CODE REVIEW

File: templates/tfengine/components/project/bigquery_datasets/main.tf [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  dataset_id = "{{.DATASET_ID}}"
+  project_id = var.project_id
+  location   = "{{get . "LOCATION" $.BIGQUERY_DATASET_LOCATION}}"
+
+  {{- if has . "DEFAULT_TABLE_EXPIRATION_MS"}}
+  default_table_expiration_ms = {{.DEFAULT_TABLE_EXPIRATION_MS}}
+  {{- end}}
+
+  {{if has . "ACCESS"}}
```
Comment (MartinPetkov) [inline templates/tfengine/components/project/bigquery_datasets/main.tf:28]: Mentioned it offline before, but I think this block can be simplified to:
```
{{hcl_has . "ACCESS" "access"}}
```
Where `hcl_has` would either return `"<field> = hclencoder.Encode(v)"` or `""` (nothing) if `has` fails.

And same with other blocks.

Comment (umairidris) [reply templates/tfengine/components/project/bigquery_datasets/main.tf]: Done

File: internal/template/funcmap.go [hunk 2]
```diff
@@ -58,6 +62,23 @@ func has(m map[string]interface{}, key string) bool {
 	return get(m, key) != nil
 }
 
+// hclField returns a hcl marshaled field e.g. `name = "foo"`
+func hclField(m map[string]interface{}, key string, req bool) (string, error) {
```
Comment (umairidris) [inline internal/template/funcmap.go:66]: Turns out you can return errors in template funcs and the template engine will handle it properly (and you don't need to handle it in the template itself). Neat.

File: internal/template/funcmap.go [hunk 3]
```diff
@@ -58,6 +62,23 @@ func has(m map[string]interface{}, key string) bool {
...
+	v, ok := m[key]
+	switch {
+	case ok:
+		b, err := hclencoder.Encode(v)
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprintf("%s = %s", key, string(b)), nil
+	case req:
```
Comment (MartinPetkov) [inline internal/template/funcmap.go:75]: This fallthrough method of handling req is pretty neat, I like it.

File: templates/tfengine/components/project/bigquery_datasets/main.tf [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+{{range get . "bigquery_datasets"}}
+module "{{resourceName .dataset_id}}" {
+  source  = "terraform-google-modules/bigquery/google"
+  version = "~> 4.2.0"
+
+  dataset_id = "{{.dataset_id}}"
+  project_id = var.project_id
+  location   = "{{get . "location" $.bigquery_dataset_location}}"
+  {{hclField . "default_table_expiration_ms" false}}
```
Comment (MartinPetkov) [inline templates/tfengine/components/project/bigquery_datasets/main.tf:23]: Should this field (`default_table_expiration_ms`) be defined in full.yaml, to show off how to use it?

Comment (umairidris) [reply templates/tfengine/components/project/bigquery_datasets/main.tf:23]: Done

File: internal/template/funcmap.go [hunk 5]
```diff
@@ -58,6 +62,23 @@ func has(m map[string]interface{}, key string) bool {
 	return get(m, key) != nil
 }
 
+// hclField returns a hcl marshaled field e.g. `name = "foo"`
+func hclField(m map[string]interface{}, key string, req bool) (string, error) {
+	v, ok := m[key]
+	switch {
+	case ok:
+		b, err := hclencoder.Encode(v)
```
Comment (xingao267) [inline internal/template/funcmap.go:70]: `b` here can be a struct as well right?

Comment (umairidris) [reply internal/template/funcmap.go:70]: b are jsut bytes, I'm guessing you mean v? v can be any primitive or list or map

Comment (xingao267) [reply internal/template/funcmap.go:70]: right

**Pull request merged**